Decrease complexity of oic.oauth2.message.Message._add_value

Part of  #311:
`src/oic/oauth2/message.py:378:1: C901 'Message._add_value' is too complex (37) [mccabe]`

I will separate the code into separate subfunctions based on type and/or refactor it a little bit.

The function is basically untested but private so I will probably try to write tests through `from_dict`.

The max-xomplexity can then be lowered to `32`.

--- a/pylama.ini
+++ b/pylama.ini
@@ -2,4 +2,4 @@
 max_line_length = 120
 
 [pylama:mccabe]
-complexity = 40
+complexity = 32

--- a/src/oic/oauth2/message.py
+++ b/src/oic/oauth2/message.py
@@ -344,90 +344,30 @@ class Message(MutableMapping):
                 continue
 
             skey = str(key)
-            try:
-                (vtyp, req, _, _deser, null_allowed) = _spec[key]
-            except KeyError:
+            if key in _spec:
+                lookup_key = key
+            else:
                 # might be a parameter with a lang tag
                 try:
                     _key, lang = skey.split("#")
                 except ValueError:
-                    try:
-                        (vtyp, _, _, _deser, null_allowed) = _spec['*']
-                        if val is None:
-                            self._dict[key] = val
-                            continue
-                    except KeyError:
-                        self._dict[key] = val
-                        continue
+                    lookup_key = '*'
                 else:
-                    try:
-                        (vtyp, req, _, _deser, null_allowed) = _spec[_key]
-                    except KeyError:
-                        try:
-                            (vtyp, _, _, _deser, null_allowed) = _spec['*']
-                            if val is None:
-                                self._dict[key] = val
-                                continue
-                        except KeyError:
-                            self._dict[key] = val
-                            continue
-
-            self._add_value(skey, vtyp, key, val, _deser, null_allowed)
+                    lookup_key = _key
+            if lookup_key in _spec:
+                (vtyp, _, _, _deser, null_allowed) = _spec[lookup_key]
+                self._add_value(skey, vtyp, key, val, _deser, null_allowed)
+            else:
+                self._dict[key] = val
         return self
 
     def _add_value(self, skey, vtyp, key, val, _deser, null_allowed):
-        # if not val:
-        # return
-
         if isinstance(val, list):
             if (len(val) == 0 or val[0] is None) and null_allowed is False:
                 return
 
         if isinstance(vtyp, list):
-            vtype = vtyp[0]
-            if isinstance(val, vtype):
-                if issubclass(vtype, Message):
-                    self._dict[skey] = [val]
-                elif _deser:
-                    try:
-                        self._dict[skey] = _deser(val, sformat="urlencoded")
-                    except Exception as exc:
-                        raise DecodeError(ERRTXT % (key, exc))
-                else:
-                    setattr(self, skey, [val])
-            elif isinstance(val, list):
-                if _deser:
-                    try:
-                        val = _deser(val, sformat="dict")
-                    except Exception as exc:
-                        raise DecodeError(ERRTXT % (key, exc))
-
-                if issubclass(vtype, Message):
-                    try:
-                        _val = []
-                        for v in val:
-                            _val.append(vtype(**dict([(str(x), y) for x, y
-                                                      in v.items()])))
-                        val = _val
-                    except Exception as exc:
-                        raise DecodeError(ERRTXT % (key, exc))
-                else:
-                    for v in val:
-                        if not isinstance(v, vtype):
-                            raise DecodeError(
-                                ERRTXT % (key, "type != %s (%s)" % (
-                                    vtype, type(v))))
-
-                self._dict[skey] = val
-            elif isinstance(val, dict):
-                try:
-                    val = _deser(val, sformat="dict")
-                except Exception as exc:
-                    raise DecodeError(ERRTXT % (key, exc))
-                else:
-                    self._dict[skey] = val
-            else:
-                raise DecodeError(ERRTXT % (key, "type != %s" % vtype))
+            self._add_value_list(skey, vtyp[0], key, val, _deser, null_allowed)
         else:
             if val is None:
                 self._dict[skey] = None
@@ -435,8 +375,7 @@ class Message(MutableMapping):
                 if vtyp is bool:
                     self._dict[skey] = val
                 else:
-                    raise ValueError(
-                        '"{}", wrong type of value for "{}"'.format(val, skey))
+                    raise ValueError('"{}", wrong type of value for "{}"'.format(val, skey))
             elif isinstance(val, vtyp):  # Not necessary to do anything
                 self._dict[skey] = val
             else:
@@ -449,14 +388,11 @@ class Message(MutableMapping):
                     try:
                         self._dict[skey] = int(val)
                     except (ValueError, TypeError):
-                        raise ValueError(
-                            '"{}", wrong type of value for "{}"'.format(val,
-                                                                        skey))
+                        raise ValueError('"{}", wrong type of value for "{}"'.format(val, skey))
                     else:
                         return
                 elif vtyp is bool:
-                    raise ValueError(
-                        '"{}", wrong type of value for "{}"'.format(val, skey))
+                    raise ValueError('"{}", wrong type of value for "{}"'.format(val, skey))
 
                 if isinstance(val, six.string_types):
                     self._dict[skey] = val
@@ -469,6 +405,59 @@ class Message(MutableMapping):
                         raise TooManyValues(key)
                 else:
                     self._dict[skey] = val
+
+    def _add_value_list(self, skey, vtype, key, val, _deser, null_allowed):
+        """Add value with internal type (``vtype``) of ``list`` to the message object.
+
+        :param skey: String representation of key
+        :param vtype: Type of object in list
+        :param key: Key for the object
+        :param val: Value of the object
+        :param _deser: Deserialization method
+        :param null_allowed: If null value is allowed
+        """
+        if isinstance(val, vtype):
+            if issubclass(vtype, Message):
+                self._dict[skey] = [val]
+            elif _deser is not None:
+                try:
+                    self._dict[skey] = _deser(val, sformat="urlencoded")
+                except Exception as exc:
+                    raise DecodeError(ERRTXT % (key, exc))
+            else:
+                setattr(self, skey, [val])
+            return
+        if isinstance(val, list):
+            if _deser is not None:
+                try:
+                    val = _deser(val, sformat="dict")
+                except Exception as exc:
+                    raise DecodeError(ERRTXT % (key, exc))
+
+            if issubclass(vtype, Message):
+                try:
+                    _val = []
+                    for v in val:
+                        _val.append(vtype(**dict([(str(x), y) for x, y in v.items()])))
+                    val = _val
+                except Exception as exc:
+                    raise DecodeError(ERRTXT % (key, exc))
+            else:
+                for v in val:
+                    if not isinstance(v, vtype):
+                        raise DecodeError(ERRTXT % (key, "type != %s (%s)" % (vtype, type(v))))
+            self._dict[skey] = val
+            return
+        if isinstance(val, dict):
+            try:
+                val = _deser(val, sformat="dict")
+            except Exception as exc:
+                raise DecodeError(ERRTXT % (key, exc))
+            else:
+                self._dict[skey] = val
+                return
+
+        raise DecodeError(ERRTXT % (key, "type != %s" % vtype))
 
     def to_json(self, lev=0, indent=None):
         if lev:

--- a/tests/test_oauth2_message.py
+++ b/tests/test_oauth2_message.py
@@ -120,6 +120,14 @@ class DummyMessage(Message):
     }
 
 
+class StarMessage(Message):
+    c_param = {'*': SINGLE_REQUIRED_STRING}
+
+
+class MessageListMessage(Message):
+    c_param = {'opt_message_list': ([Message], False, None, None, False)}
+
+
 class TestMessage(object):
     def test_json_serialization(self):
         item = DummyMessage(req_str="Fair", opt_str="game", opt_int=9,
@@ -216,6 +224,72 @@ class TestMessage(object):
         assert cls.get("opt_int", 8) == 9
         assert cls.get("missing") is None
         assert cls.get("missing", []) == []
+
+    def test_from_dict_simple(self):
+        _dict = {'req_str': 'Fair', 'opt_str': 'game', 'opt_int': 9}
+
+        message = DummyMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_lang(self):
+        _dict = {'req_str#en': 'Fair', 'opt_str': 'game', 'opt_int': 9}
+
+        message = DummyMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_wrong_lang_key_no_star(self):
+        _dict = {'bad_str#en': None}
+
+        message = DummyMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_wrong_lang_key_star(self):
+        _dict = {'bad_str#en': 'test'}
+
+        message = StarMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_wrong_lang_key_star_none(self):
+        _dict = {'bad_str#en': None}
+
+        message = StarMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_wrong_key_no_star(self):
+        _dict = {'req_str_en': None}
+
+        message = DummyMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_wrong_key_star(self):
+        _dict = {'req_str_en': 'test'}
+
+        message = StarMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_wrong_key_star_none(self):
+        _dict = {'req_str_en': None}
+
+        message = StarMessage().from_dict(_dict)
+        assert message.to_dict() == _dict
+
+    def test_from_dict_empty_val(self):
+        _dict = {'req_str': '', 'req_str_list': ['']}
+
+        message = DummyMessage().from_dict(_dict)
+        assert message.to_dict() == {}
+
+    def test_from_dict_message(self):
+        _dict = {'opt_message_list': DummyMessage(req_str='test')}
+
+        message = MessageListMessage().from_dict(_dict)
+        assert message.to_dict() == {'opt_message_list': [{'req_str': 'test'}]}
+
+    def test_from_dict_message_list(self):
+        _dict = {'opt_message_list': [DummyMessage(req_str='test')]}
+
+        message = MessageListMessage().from_dict(_dict)
+        assert message.to_dict() == {'opt_message_list': [{'req_str': 'test'}]}
 
 
 class TestAuthorizationRequest(object):

